ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -8595,6 +8634,94 @@ These props are now enabled elsewhere (see commit diffs).  It would be better in
     SessionState.get().getConf().set(AcidUtils.CONF_ACID_KEY, "true");
   }
 
+  private Operator genConversionSelectOperatorByAddPartition(String dest, QB qb, Operator input,
+      Deserializer deserializer, Table table, Map<String, String> partitionSpec) throws SemanticException {
+    StructObjectInspector oi = null;
+    try {
+      oi = (StructObjectInspector) deserializer.getObjectInspector();
+    } catch (Exception e) {
+      throw new SemanticException(e);
+    }
+
+    // Check column number
+    List<? extends StructField> tableFields = oi.getAllStructFieldRefs();
+    List<ColumnInfo> rowFields = opParseCtx.get(input).getRowResolver().getColumnInfos();
+    int inColumnCnt = rowFields.size();
+    int outColumnCnt = tableFields.size();
+    List<ExprNodeDesc> expressions = new ArrayList<>(outColumnCnt);
+
+    // if target table is always unpartitioned, then the output object inspector will already contain the partition cols
+    // too, therefore we shouldn't add the partition col num to the output col num
+    boolean alreadyContainsPartCols = Optional.ofNullable(table)
+        .map(Table::getStorageHandler)
+        .map(HiveStorageHandler::alwaysUnpartitioned)
+        .orElse(Boolean.FALSE);
+
+    AtomicBoolean convert = new AtomicBoolean(false);
+
+    if (inColumnCnt < outColumnCnt && alreadyContainsPartCols && partitionSpec != null) {
```

Comment:
Could you please invert this if?
```
if (!(inColumnCnt < outColumnCnt && alreadyContainsPartCols && partitionSpec != null)) {
  return input;
}
int rowNum = 0;
...
```
It reduces indentation and improve readability.